### PR TITLE
Rewrite CommandChannel

### PR DESCRIPTION
**Pull request by VirrageS**

### CODE REVIEW

File: test/run_test.sh [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env bash
 set -e
 
-PYCMD=${PYCMD:="python"}
```
Comment (apaszke) [inline test/run_test.sh]: nope

File: torch/lib/THD/base/channels/ChannelUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+  struct sockaddr_in addr;
+  socklen_t addr_len = sizeof(addr);
+  SYSCHECK(::getsockname(socket, reinterpret_cast<struct sockaddr*>(&addr), &addr_len))
+  std::uint16_t listen_port = ntohs(addr.sin_port);
+
+  return std::make_tuple(socket, listen_port);
+}
+
+
+int connect(const std::string& address, std::uint16_t port, int wait) {
```
Comment (apaszke) [inline torch/lib/THD/base/channels/ChannelUtils.cpp]: `bool wait`?

File: torch/lib/THD/base/channels/ChannelUtils.hpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+inline std::uint16_t convertToPort(long port) {
+  if ((port < 0) || (port >= UINT16_MAX))
+    throw std::domain_error("invalid port");
+
+  return static_cast<std::uint16_t>(port);
+}
+
+inline std::uint32_t convertToRank(long rank, long min = 0) {
+  if ((rank < min) || (rank >= UINT32_MAX))
+    throw std::domain_error("invalid rank");
```
Comment (apaszke) [inline torch/lib/THD/base/channels/ChannelUtils.hpp]: add sth like `(value of out range)` here and above

File: torch/lib/THD/base/channels/ChannelUtils.hpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+namespace thd {
+
+#define SYSCHECK(expr) { \
+  errno = 0; (expr);     \
+  if (errno != 0) throw std::system_error(errno, std::system_category()); \
+}
+
+template<typename T>
+void send_bytes(int socket, const T* buffer, std::size_t length)
```
Comment (apaszke) [inline torch/lib/THD/base/channels/ChannelUtils.hpp]: Why did you remove that `MSG_MORE` optimization? Either keep it, or even better add a specialization for vector writes.

File: torch/lib/THD/base/channels/command_channel/CommandChannel.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+MasterCommandChannel::~MasterCommandChannel() {
+  for (auto socket : _sockets) {
+    if (socket != -1)
+      ::close(socket);
+  }
+}
+
+bool MasterCommandChannel::init() {
+  // listen on workers
```
Comment (apaszke) [inline torch/lib/THD/base/channels/command_channel/CommandChannel.cpp]: This comment seems unnecessary

File: torch/lib/THD/base/channels/command_channel/CommandChannel.cpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  std::uint32_t rank;
+  for (std::size_t i = 1; i < _sockets.size(); ++i) {
+    std::tie(socket, std::ignore) = accept(_sockets[0]);
+    recv_bytes<std::uint32_t>(socket, &rank, 1);
+    _sockets.at(rank) = socket;
+  }
+
+  /* Sending confirm byte is necessary to block workers until all remaing workers
+   * connect. This necessesity comes from case when worker finishes connecting
+   * to command channel and start connecting to data channel. Since master
```
Comment (apaszke) [inline torch/lib/THD/base/channels/command_channel/CommandChannel.cpp]: I thought we always connect the data channel before we initialize the command channel. Isn't it right?

File: torch/lib/THD/base/channels/command_channel/CommandChannel.cpp [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+  }
+
+   // close listen socket
+  ::close(_sockets[0]);
+  _sockets[0] = -1;
+  return true;
+}
+
+void MasterCommandChannel::sendMessage(std::unique_ptr<rpc::RPCMessage> msg, int rank) {
+  if (rank <= 0) {
```
Comment (apaszke) [inline torch/lib/THD/base/channels/command_channel/CommandChannel.cpp]: check world size?

File: torch/lib/THD/base/channels/command_channel/CommandChannel.cpp [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    throw std::domain_error("sendMessage received invalid rank as parameter");
+  }
+
+  ::thd::sendMessage(_sockets.at(rank), std::move(msg));
+}
+
+std::unique_ptr<rpc::RPCMessage> MasterCommandChannel::recvMessage(int rank) {
+  if (rank <= 0) {
```
Comment (apaszke) [inline torch/lib/THD/base/channels/command_channel/CommandChannel.cpp]: check world size?

File: torch/lib/THD/master_worker/common/ByteArray.cpp [hunk 9]
```diff
@@ -11,111 +11,52 @@
...
+    _data.reserve(_data.capacity() + size);
+  }
 
-ByteArray ByteArray::fromData(const char* arr, std::size_t size) {
-  char* new_arr = static_cast<char*>(std::malloc(size));
-  if (new_arr == nullptr) {
-    throw std::system_error(errno,
-                            std::system_category(),
-                            "failed to allocate memory");
+  for (std::size_t i = 0; i < size; ++i) {
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/ByteArray.cpp]: Why not `_data.append(arr, arr + size)`?

File: torch/lib/THD/master_worker/common/ByteArray.cpp [hunk 10]
```diff
@@ -11,111 +11,52 @@
...
-}
-
-std::size_t ByteArray::length() const {
-  return _length;
-}
-
-std::string ByteArray::to_string() {
-  return std::string(_data, _length);
-}
+  if (_data.capacity() < _data.size() + size) {
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/ByteArray.cpp]: I don't think this is really necessary. `reserve` never makes the block smaller, and `append` should take care of resizing the string appropriately.

File: torch/lib/THD/master_worker/common/ByteArray.cpp [hunk 11]
```diff
@@ -11,111 +11,52 @@
...
+std::size_t ByteArray::length() const {
+  return _data.size();
 }
 
-void ByteArray::_resizeExact(std::size_t desired_size) {
-  if (desired_size <= _size)
-    return;
-  _realloc(desired_size);
+std::string ByteArray::to_string() const {
+  return std::string(_data.begin(), _data.end());
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/ByteArray.cpp]: I think `return _data` should do it. It's already a string, and since you return by value, it will make a copy automatically.

**Review by apaszke — COMMENTED** [review]
You checked in an additional binary file (test/data/legacy_serialized.py).

Also, we're not going to be implementing any more CommandChannels, so I think it makes sense to keep the directory structure as it was. Just put ChannelUtils.cpp in `base/` directly and keep only the data channels in `channels/` (maybe it makes sense to rename it to `data_channels`).

File: torch/lib/THD/base/channels/ChannelUtils.hpp (hunk 4 above)
Comment (VirrageS) [reply torch/lib/THD/base/channels/ChannelUtils.hpp]: I didn't. It is just in another PR. Should I add it here to?

File: torch/lib/THD/base/channels/command_channel/CommandChannel.cpp (hunk 6 above)
Comment (VirrageS) [reply torch/lib/THD/base/channels/command_channel/CommandChannel.cpp]: Oh, okay. I thought command channel the data channel. Should I keep that anyway or remove it?

File: torch/lib/THD/master_worker/common/ByteArray.cpp (hunk 11 above)
Comment (VirrageS) [reply torch/lib/THD/master_worker/common/ByteArray.cpp]: `_data` is vector of chars. Should I change type to `std::string`?

Comment (VirrageS) [reply torch/lib/THD/master_worker/common/ByteArray.cpp]: Nvm, I will change that to `std::string` it make more sense.

File: torch/lib/THD/base/channels/command_channel/CommandChannel.cpp (hunk 6 above)
Comment (VirrageS) [reply torch/lib/THD/base/channels/command_channel/CommandChannel.cpp]: I decided to keep that but change description. I think that even this can cause some small performance downgrade, it is a good idea to 'test' the connection and barrier workers. It also won't cause any problems in future if we decide to change order of channels initialization.

Comment (apaszke) [reply torch/lib/THD/base/channels/command_channel/CommandChannel.cpp]: Sure. It's just that incorrect comments are worse than no comments :wink:
